# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h
# Commit: 8d80a4ba914c
# Full Hash: 8d80a4ba914c6f81430ba1c55471a8d2b3aac60e
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Create CallWorkerThread which implements AbstractThread for a TaskQueueWrapper. r=ng
#   
#   Ideally TaskQueueWrapper would implement both webrtc::TaskQueueBase and
#   AbstractThread, but webrtc::TaskQueueBase is not refcounted so implementing both
#   in the same class is not possible.
# ==============================================================================

diff -r e0a2e54e4de8 -r 8d80a4ba914c dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h
--- a/dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h	Thu Apr 08 11:31:43 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h	Mon Apr 12 14:49:20 2021 +0200
@@ -17,6 +17,7 @@
 
 namespace mozilla {
 class MediaSessionConduit;
+class SharedWebrtcState;
 
 // Wrap the webrtc.org Call class adding mozilla add/ref support.
 class WebrtcCallWrapper {
@@ -26,12 +27,14 @@
   static RefPtr<WebrtcCallWrapper> Create(
       const dom::RTCStatsTimestampMaker& aTimestampMaker,
       UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
-      SharedWebrtcState* aSharedState, webrtc::WebRtcKeyValueConfig* aTrials);
+      const RefPtr<SharedWebrtcState>& aSharedState,
+      webrtc::WebRtcKeyValueConfig* aTrials);
 
   static RefPtr<WebrtcCallWrapper> Create(
       const dom::RTCStatsTimestampMaker& aTimestampMaker,
       UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
-      webrtc::SharedModuleThread* aModuleThread,
+      RefPtr<AbstractThread> aCallThread,
+      rtc::scoped_refptr<webrtc::SharedModuleThread> aModuleThread,
       const webrtc::AudioState::Config& aAudioStateConfig,
       webrtc::AudioDecoderFactory* aAudioDecoderFactory,
       webrtc::WebRtcKeyValueConfig* aTrials);
@@ -44,6 +47,8 @@
   WebrtcCallWrapper(const WebrtcCallWrapper&) = delete;
   void operator=(const WebrtcCallWrapper&) = delete;
 
+  void SetCall(UniquePtr<webrtc::Call> aCall);
+
   webrtc::Call* Call() const;
 
   bool UnsetRemoteSSRC(uint32_t ssrc);
@@ -81,10 +86,10 @@
   virtual ~WebrtcCallWrapper();
 
  private:
-  WebrtcCallWrapper(RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
+  WebrtcCallWrapper(RefPtr<AbstractThread> aCallThread,
+                    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
                     UniquePtr<webrtc::VideoBitrateAllocatorFactory>
                         aVideoBitrateAllocatorFactory,
-                    UniquePtr<webrtc::Call> aCall,
                     UniquePtr<webrtc::RtcEventLog> aEventLog,
                     UniquePtr<webrtc::TaskQueueFactory> aTaskQueueFactory,
                     const dom::RTCStatsTimestampMaker& aTimestampMaker,
@@ -99,6 +104,7 @@
   UniquePtr<media::ShutdownBlockingTicket> mShutdownTicket;
 
  public:
+  const RefPtr<AbstractThread> mCallThread;
   const RefPtr<webrtc::AudioDecoderFactory> mAudioDecoderFactory;
   const UniquePtr<webrtc::VideoBitrateAllocatorFactory>
       mVideoBitrateAllocatorFactory;
@@ -106,7 +112,7 @@
   const UniquePtr<webrtc::TaskQueueFactory> mTaskQueueFactory;
 
  private:
-  // Main thread only, as it's the Call worker thread.
+  // Call worker thread only.
   UniquePtr<webrtc::Call> mCall;
 };
 